### PR TITLE
Feature/3d basis evaluate cherry

### PR DESCRIPTION
**Pull request by will-saunders-ukaea**
# 3D Evaluation and Projection

* Added eModified_C, eModified_PyrC reference and kernel implementations
* Added reference implementations of basis functions and looping for 3D elements
* Added CRTP abstract interface for loops that evaluating all modes for Hexes, Prisms, Pyramids, Tets, Triangles and Quads.
* Moved projection and evaluation over to CRTP templated loops and kernels, 2D and 3D.
* Tests for the above.

### CODE REVIEW

**Comment by jwscook** [comment]
@jwscook do this!

**Comment by jwscook** [comment]
gcc-hipsycl build for CPU:

```
[ RUN      ] ParticleFunctionEvaluation3D.ContField
Fatal   : Level 1 assertion violation
BwdTrans3DWorkspace: Requires homogenous points.
unknown file: Failure
C++ exception with description "Level 1 assertion violation
BwdTrans3DWorkspace: Requires homogenous points." thrown in the test body.
[  FAILED  ] ParticleFunctionEvaluation3D.ContField (13392 ms)
```

```
[ RUN      ] ParticleFunctionProjection3DBasisEval.ContField
Fatal   : Level 1 assertion violation
BwdTrans3DWorkspace: Requires homogenous points.
unknown file: Failure
C++ exception with description "Level 1 assertion violation
BwdTrans3DWorkspace: Requires homogenous points." thrown in the test body.
[  FAILED  ] ParticleFunctionProjection3DBasisEval.ContField (4086 ms)
```

```
[ RUN      ] ParticleFunctionBasisEvaluation.Basis3D
Fatal   : Level 1 assertion violation
BwdTrans3DWorkspace: Requires homogenous points.
unknown file: Failure
C++ exception with description "Level 1 assertion violation
BwdTrans3DWorkspace: Requires homogenous points." thrown in the test body.
[  FAILED  ] ParticleFunctionBasisEvaluation.Basis3D (653 ms)
```

**Comment by jwscook** [comment]
Same for the oneapi build

**Comment by jwscook** [comment]
After a rebuild this problem went away.

**Comment by jwscook** [comment]
The unit and integration tests pass for CPU and my nvidia card on my desktop.

File: src/nektar_interface/basis_reference.cpp [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+ *  @returns Basis function evaluated at point.
+ */
+double eval_modB_ij(const int p, const int q, const double z) {
+
+  double output;
+
+  if (p == 0) {
+    output = eval_modA_i(q, z);
+  } else if (q == 0) {
+    output = std::pow(0.5 * (1.0 - z), (double)p);
```
Comment (jwscook) [inline src/nektar_interface/basis_reference.cpp:39]: Converting int p to a double in pow will make it slower, unless that's outdated info.

File: src/nektar_interface/basis_reference.cpp [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+  } else {
+    output = std::pow(0.5 * (1.0 - z), (double)p) * 0.5 * (1.0 + z) *
```
Comment (jwscook) [inline src/nektar_interface/basis_reference.cpp:41]: ditto

File: test/integration/nektar_interface/test_function_projection_order.cpp [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+#include <gtest/gtest.h>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <random>
+#include <set>
+#include <string>
+#include <vector>
+
+using namespace std;
```
Comment (jwscook) [inline test/integration/nektar_interface/test_function_projection_order.cpp:21]: Can this line be deleted?

**Review by jwscook — APPROVED** [review]

File: include/nektar_interface/function_basis_evaluation.hpp [hunk 4]
```diff
@@ -158,43 +116,52 @@ class FunctionEvaluateBasis : public BasisEvaluateBase<T> {
...
             const INT cellx = k_cells_iterset[iter_cell];
             const INT layerx = idx.get_global_id(1);
+            ExpansionLooping::JacobiExpansionLoopingInterface<EVALUATE_TYPE>
+                loop_type{};
 
             if (layerx < d_npart_cell[cellx]) {
-              const double *dofs = &k_global_coeffs[k_coeffs_offsets[cellx]];
+              const REAL *dofs = &k_global_coeffs[k_coeffs_offsets[cellx]];
 
               // Get the number of modes in x and y
```
[the remark this answers is not in the document]
Comment (will-saunders-ukaea) [reply include/nektar_interface/function_basis_evaluation.hpp]: done

File: include/nektar_interface/function_basis_evaluation.hpp [hunk 5]
```diff
@@ -158,43 +116,52 @@ class FunctionEvaluateBasis : public BasisEvaluateBase<T> {
...
+                xi2 = k_ref_positions[cellx][2][layerx];
+              }
 
-              GeometryInterface::BaseCoordinateMapping2D<MAP_TYPE>
-                  coord_mapping{};
-              coord_mapping.loc_coord_to_loc_collapsed(xi0, xi1, &eta0, &eta1);
+              loop_type.loc_coord_to_loc_collapsed(xi0, xi1, xi2, &eta0, &eta1,
+                                                   &eta2);
 
               // Get the local space for the 1D evaluations in dim0 and dim1
```
[the remark this answers is not in the document]
Comment (will-saunders-ukaea) [reply include/nektar_interface/function_basis_evaluation.hpp]: done

File: include/nektar_interface/function_basis_evaluation.hpp [hunk 6]
```diff
@@ -158,43 +116,52 @@ class FunctionEvaluateBasis : public BasisEvaluateBase<T> {
...
+              REAL *local_space_0 =
                   &local_mem[idx_local *
-                             (k_max_total_nummodes0 + k_max_total_nummodes1)];
-              double *local_space_1 = local_space_0 + k_max_total_nummodes0;
+                             (k_max_total_nummodes0 + k_max_total_nummodes1 +
+                              k_max_total_nummodes2)];
+              REAL *local_space_1 = local_space_0 + k_max_total_nummodes0;
+              REAL *local_space_2 = local_space_1 + k_max_total_nummodes1;
 
               // Compute the basis functions in dim0 and dim1
```
[the remark this answers is not in the document]
Comment (will-saunders-ukaea) [reply include/nektar_interface/function_basis_evaluation.hpp]: done

File: src/nektar_interface/basis_reference.cpp (hunk 1 above)
Comment (will-saunders-ukaea) [reply src/nektar_interface/basis_reference.cpp]: done

File: src/nektar_interface/basis_reference.cpp (hunk 2 above)
Comment (will-saunders-ukaea) [reply src/nektar_interface/basis_reference.cpp]: done

File: test/integration/nektar_interface/test_function_projection_order.cpp (hunk 3 above)
Comment (will-saunders-ukaea) [reply test/integration/nektar_interface/test_function_projection_order.cpp]: done

**Pull request merged**